File: C++/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *cycleFptr(ListNode *head) {  // if cycle found, returns the fast pointer 
        
        if (head != NULL and head->next != NULL) {
            ListNode* fptr = head;
            ListNode* sptr = head;

            while (fptr != NULL and fptr->next != NULL) {
                fptr = fptr->next->next;
                sptr = sptr->next;

                if (fptr == sptr){
                    return fptr;
                }
            }      
        }
        return NULL;  // no cycle found           
    }

    ListNode *detectCycle(ListNode *head) {
        ListNode* fptr = cycleFptr(head);
        if (fptr){
            ListNode* sptr = head;
            
            while (fptr != sptr) {
                fptr = fptr->next;
                sptr = sptr->next;
            }
        }            
        return fptr;
    }
};

File: LeetCode/Python/linkedlist-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def cycleFptr(self, head: ListNode) -> ListNode:
        if head and head.next:
            fptr, sptr = head, head
            
            while fptr and fptr.next:
                fptr, sptr = fptr.next.next, sptr.next
                
                if fptr == sptr:
                    return fptr
        
        return None
        
        
    def detectCycle(self, head: ListNode) -> ListNode:
        fptr = self.cycleFptr(head)
        
        if fptr:
            sptr = head
            
            while fptr is not sptr:
                fptr, sptr = fptr.next, sptr.next
                
        return fptr
    
